bindings/pyroot/pythonizations/python/ROOT/_pythonization/_tcontext.py

```diff
@@ -0,0 +1,101 @@
+
+################################################################################
+# Copyright (C) 1995-2022, Rene Brun and Fons Rademakers.                      #
+# All rights reserved.                                                         #
+#                                                                              #
+# For the licensing terms see $ROOTSYS/LICENSE.                                #
+# For the list of contributors see $ROOTSYS/README/CREDITS.                    #
+################################################################################
+
+r'''
+/**
+\class TDirectory::TContext
+\brief \parblock \endparblock
+\htmlonly
+<div class="pyrootbox">
+\endhtmlonly
+## PyROOT
+
+The functionality offered by TContext can be used in PyROOT with a context manager. Here are a few examples:
+\code{.py}
+import ROOT
+from ROOT import TDirectory
+
+with TDirectory.TContext():
+    # Open some file here
+    file = ROOT.TFile(...)
+    # Retrieve contents from the file
+    histo = file.Get("myhisto")
+
+# After the 'with' statement, the current directory is restored to ROOT.gROOT
+\endcode
+\n 
+\code{.py}
+import ROOT
+from ROOT import TDirectory
+
+file1 = ROOT.TFile("file1.root", "recreate")
+#...
+file2 = ROOT.TFile("file2.root", "recreate")
+#...
+file3 = ROOT.TFile("file3.root", "recreate")
+
+# Before the 'with' statement, the current directory is file3 (the last file opened)
+with TDirectory.TContext(file1):
+    # Inside the statement, the current directory is file1
+    histo = ROOT.TH1F(...)
+    histo.Write()
+
+# After the statement, the current directory is restored to file3
+\endcode
+\n 
+\code{.py}
+import ROOT
+from ROOT import TDirectory
+
+file1 = ROOT.TFile("file1.root")
+file2 = ROOT.TFile("file2.root")
+
+with TDirectory.TContext(file1, file2):
+    # Manage content in file2
+    histo = ROOT.TH1F(...)
+    # histo will be written to file2
+    histo.Write()
+
+# Current directory is set to 'file1.root'
+\endcode
+
+Note that TContext restores the current directory to its status before the 'with'
+statement, but does not change the status of any file that has been opened inside
+the context (e.g. it does not automatically close the file).
+\htmlonly
+</div>
+\endhtmlonly
+*/
+'''
+
+from . import pythonization
+
+
+def _TContextExit(ctxt, exc_type, exc_val, exc_tb):
+    """
+    Destroy the TContext, mimicking its RAII behaviour.
+    Signature and return value are imposed by Python, see
+    https://docs.python.org/3/library/stdtypes.html#typecontextmanager.
+    """
+    ctxt.__destruct__()
+    return False
+
+
+@pythonization("TContext", ns="TDirectory")
+def pythonize_tcontext(klass):
+    """
+    TContext pythonizations:
+    - __enter__ and __exit__ methods are defined to make it work as a context manager.
+    """
+
+    # The user doesn't need to do anything with the TContext object itself, so
+    # we don't provide it in the context manager expression
+    klass.__enter__ = lambda ctxt: None
```

Comment:
Doesn't this mean that the context object gets immediately destroyed? It wouldn't hurt perhaps to still return it?

Reply:
The [docs of the `with` statement](https://docs.python.org/3/reference/compound_stmts.html#the-with-statement) make me think that the object is kept alive by Python even when not explicitly returned to the user, for example with sentences like this

> The context manager’s __enter__() is loaded for later use.

I had a doubt that this was working with TContext only because the context registers itself with the directory data member `fDirectory`, so I thought that maybe it was being kept alive by that only and not by Python. So I tried this example with a TH1F

```python
import ROOT
from ROOT import pythonization


def _enter(h):
    print("Inside __enter__ function")
    h.SetDirectory(ROOT.nullptr)
    print(f"{h.GetDirectory()=}")
    print(f"{h=}")
    return None

def _exit(h, exc_type, exc_value, traceback):
    print("Inside __exit__ function")
    print(f"{h=}")
    print(f"{h.GetEntries()=}")
    return False

@pythonization("TH1F")
def pythonize(klass):
    klass.__enter__ = _enter
    klass.__exit__ = _exit

if __name__ == "__main__":
    with ROOT.TH1F("myname","mytitle", 10, 0, 10):
        print("In context")
```

And the output is
```
Inside __enter__ function
h.GetDirectory()=<cppyy.gbl.TDirectory object at 0x(nil)>
h=<cppyy.gbl.TH1F object at 0x56117a2bc870>
In context
Inside __exit__ function
h=<cppyy.gbl.TH1F object at 0x56117a2bc870>
h.GetEntries()=0.0
```

Here the `__enter__` function returns `None`, but the `__exit__` function is still able to access the same histogram which is still alive. So I think we can take this as Python keeping the object alive until the end of the statement.

In general, since TContext has no public API, I think it would be better to not pass the object to the user since they are not supposed to do anything with it.

Reply:
Ok, all seems good then!